orttraining/orttraining/core/graph/pattern_graph/pattern_node.h

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/graph/contrib_ops/onnx_function_util.h"
+#include "orttraining/core/graph/graph_augmenter.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/model.h"
+
+namespace onnxruntime {
+namespace training {
+
+/**
+ * @brief Pattern graph input data types.
+ * Class representing allowed data types for inputs of the pattern graph.
+ */
+struct PGraphInputTypes {
+  friend class PGraphInput;
+
+ public:
+  /**
+    @brief Predefined categories for common used data type set.
+  */
+  enum class TypesCategory {
+    AllIntegerTensorTypes,
+    AllFloatTensorTypes,
+  };
+
+ public:
+  /* Construct allowed data types from predefined categories. */
+  PGraphInputTypes(TypesCategory category);
+  /* Construct allowed data types from explicit list of data types. */
+  PGraphInputTypes(const std::vector<ONNX_NAMESPACE::TensorProto_DataType>& allowed_types)
+      : allowed_types_(allowed_types) {}
+
+ private:
+  /**
+   * We need a valid data type for the input when building the pattern graph.
+   * Here we choose the first data type in the allowed list. For user of this class,
+   * default type is meaningless, and will not be used as strict type checking,
+   * instead, if target input/output arg's data type in allowed_types_, then it is a match.
+   */
+  ONNX_NAMESPACE::TensorProto_DataType GetDefaultType() const {
+    ORT_ENFORCE(allowed_types_.size(), "Empty type list in PGraphInputTypes.");
+    return allowed_types_.at(0);
+  }
+
+  void Init(const std::vector<ONNX_NAMESPACE::TensorProto_DataType>& types) {
+    allowed_types_ = types;
+  }
+
+  std::vector<ONNX_NAMESPACE::TensorProto_DataType> allowed_types_;
+};
+
+/**
+ * @brief Pattern graph input description.
+ * Class representing a graph input to a pattern graph. Two kinds of graph inputs are supported:
+ * > Graph inputs, which is generated from other subgraphs.
+ * > Constant node, which consume no else, but generate a constant value.
+ *
+ * During node matching, by default, all user given fields will be checked, compared with
+ * the target graph.
+ */
+struct PGraphInput {
+  friend class PatternGraph;
+
+ public:
+  /*
+   * is_dangling: if it's set to false, this arg would not be required  to match an arg in target graph
+   * is_constant: it indicates if the arg is a constant.
+   * Since we reuse Augumenter to implement the pattern graph, we cannot get real constant but only use a flag.
```

Comment:
In ```Augumenter``` we set the ```optype``` of ```NodeDef``` as "constant" to get a constant, it seems that we cannot get a variable initializer instead of constant.

Reply:
did not get your point, what is the problem here?

Reply:
There's no problem with the code but just a detail. Maybe the comment above should be "we cannot get real non-constant initializer but only use a flag".